Made a Geolocation extension

samuellouf opened the pull request:
Hello, folks!
I made a Geolocation extension.
Enjoy!

samuellouf committed:
Fixed Geolocation extension

yuri-kiss commented:
This is pretty simple, I will give it a review later today if I don't forget

samuellouf commented:
> This is pretty simple, I will give it a review later today if I don't forget

Thanks!

samuellouf committed:
Merge branch 'TurboWarp:master' into Geolocation

samuellouf commented:
Okay, I'm just gonna add Geocoding Blocks.

samuellouf committed:
Merge branch 'TurboWarp:master' into Geolocation

samuellouf committed:
Added Translations to Menus

samuellouf commented:
Here ! It's finally done!

samuellouf committed:
Merge branch 'TurboWarp:master' into Geolocation

samuellouf committed:
Merge branch 'TurboWarp:master' into Geolocation

Brackets-Coder commented on extensions/SamuelLouf/Geolocation.js:
```diff
@@ -0,0 +1,198 @@
+  "use strict";
+
+  function getGeolocation(
+    options = { enableHighAccuracy: true, timeout: 5000, maximumAge: 0 }
```
I'm not familiar with geolocation apis but will the timeout affect anything?

yuri-kiss replied:
it will slow it down iirc

samuellouf replied:
Not really : timeout is the limit you give to geolocate the device.
In this case, the `getGeolocation` function gives the browser 5000ms (or 5s) tops to geolocate the device.
It's the `setTimeout` function that slows down/stops the code. See on [mdn web docs](https://developer.mozilla.org/en-US/docs/Web/API/Window/setTimeout).

Brackets-Coder commented on extensions/SamuelLouf/Geolocation.js:
```diff
@@ -0,0 +1,198 @@
+      };
+    }
+
+    async getCurrent(args) {
+      if (!(await this.isAllowed())) return "";
+      var coordinates = await getGeolocation();
+      if (coordinates.success == true) {
+        return coordinates[args.WHAT];
+      } else {
+        return "";
+      }
+    }
+
+    async isAllowed() {
+      return await Scratch.canGeolocate();
+    }
```
@yuri-kiss are we allowed to make block functions themselves asynchronous or should we put asynchronous functions inside them?

yuri-kiss replied:
they can be async

Brackets-Coder reviewed (commented):
Looks fine to me! All protocols are being followed and I don't see any reason why this can't be merged now.

Brackets-Coder commented:
> I can't find any blocks to actually get my location, which is a pretty necessary feature for a geolocation extension.

huh didn't notice that, maybe we're missing something?

CST1229 commented:
> I can't find any blocks to actually get my location, which is a pretty necessary feature for a geolocation extension.

it's an error with the extension. the block has the wrong menu (it should be `coordinates`)
<img width="426" height="208" alt="image" src="https://github.com/user-attachments/assets/7155f45c-da08-44fe-972d-b8a42c04e89c" />

samuellouf committed:
Update Geolocation.js

samuellouf committed:
Merge branch 'TurboWarp:master' into Geolocation

samuellouf commented:
Fixed it!!

PPPDUD reviewed (approved):
I am quite impressed by this extension. All of the relevant blocks are there, and all checks have passed. @samuellouf You have my blessing.

Brackets-Coder reviewed (changes requested):
Get longitude and latitude doesn't seem to work for me in Safari (the block says it's supported and it's not an API limitation), and in Firefox longitude is negative where for me it should be positive according to my Maps app.

Once these issues are fixed I think we'll be good to go

Brackets-Coder commented:
> Get longitude and latitude doesn't seem to work for me in Safari (the block says it's supported), and in Firefox longitude is negative where for me it should be positive according to my Maps app.

samuellouf committed:
Made a few important changes

 - Changed the default timeout
 - Fixed the `getCurrent` timeout (it wasn't taken into account until this fix)

samuellouf commented:
> Get longitude and latitude doesn't seem to work for me in Safari (the block says it's supported and it's not an API limitation), and in Firefox longitude is negative where for me it should be positive according to my Maps app.

Perhaps there's something wrong with your computer or your phone. (No offense)
It works for me on both Google Chrome and Firefox on Windows 10, 11, and Linux Mint.

samuellouf committed:
Merge branch 'master' into Geolocation

samuellouf commented:
@PPPDUD, could you check if the position the extension returns is valid?

Brackets-Coder reviewed (changes requested):
Still doesn't seem to work in Safari

<img width="1840" height="1195" alt="Image" src="https://github.com/user-attachments/assets/44c22bf0-db6c-4cb2-8cbe-9547d5f762c4" />

samuellouf commented:
> > > Get longitude and latitude doesn't seem to work for me in Safari (the block says it's supported and it's not an API limitation), and in Firefox longitude is negative where for me it should be positive according to my Maps app.
> > 
> > 
> > Perhaps there's something wrong with your computer or your phone. (No offense) It works for me on both Google Chrome and Firefox on Windows 10, 11, and Linux Mint.
> 
> It's not the computer, I'm on a modern Mac with the latest software version. It works in Firefox, not Safari- it's not a browser limitation because the extension says I support geolocation and Apple's MapJS kit wouldn't work without geolocation capabilities, so something about the get location reporter doesn't work in Safari.

That's interesting. I don't get why it isn't working. It _should_ work.
Is the longitude still negative?

samuellouf committed:
Formated it

samuellouf committed:
Merge branch 'Geolocation' of https://github.com/samuellouf/extensions into Geolocation

Brackets-Coder commented:
> > > > Get longitude and latitude doesn't seem to work for me in Safari (the block says it's supported and it's not an API limitation), and in Firefox longitude is negative where for me it should be positive according to my Maps app.
> > > 
> > > 
> > > Perhaps there's something wrong with your computer or your phone. (No offense) It works for me on both Google Chrome and Firefox on Windows 10, 11, and Linux Mint.
> > 
> > 
> > It's not the computer, I'm on a modern Mac with the latest software version. It works in Firefox, not Safari- it's not a browser limitation because the extension says I support geolocation and Apple's MapJS kit wouldn't work without geolocation capabilities, so something about the get location reporter doesn't work in Safari.
> 
> That's interesting. I don't get why it isn't working. It _should_ work. Is the longitude still negative?

Yes, but I'm not 180º on the other side of the world. A theory, I updated to the latest major macOS Tahoe release recently and it's a .0 version. I've discovered a few UI inconsistencies after the redesign so there may be potential functionality issues in the latest version, but it's very unlikely.

Brackets-Coder commented:
> Should I add a "when user moves" block? (It would be a hat, ofc) It would be paired with latitude, longitude, accuracy blocks. It seems to be easily doable.

Depends on whether users would use it and how performant it would be

samuellouf committed:
Made various changes

Added "onUserMove"

samuellouf committed:
Formated the extension

Commit:
Update Geolocation.js to improve grammar

samuellouf reviewed (commented):
My bad. English isn't my strong suit sometimes.

samuellouf committed:
Merge branch 'TurboWarp:master' into Geolocation

samuellouf committed:
Added the blocks' border color (light green)

samuellouf reviewed (commented):
Added the border

Brackets-Coder commented:
I'll have to review this again...

Brackets-Coder commented:
Give me until the weekend though please

Brackets-Coder commented:
Sorry I forgot, please remind me this Saturday. Thanks so much for your patience!

Brackets-Coder reviewed (changes requested):
- Setting the accuracy does not appear to affect the result of "get [accuracy]" which appears to always report the same value
- Longitude still appears to be negative for me in two different browsers. 😕 This can be mitigated with a simple subtract operator, but will likely cause confusion for users when the extension says they're on the other side of the world
- "When the user's position changes" hat block is always triggered when the "get [position]" reporters are executed, even if their result is the same as the last time they executed.

samuellouf commented:
Thanks for your feedback! I'll try to work on it this weekend.

samuellouf commented:
I actually took a look at some of it now.

- The "When the user's position changes" gets triggered by accuracy changes and slight changes in the position. I'm overlooking all position changes under `5*10**(-5)` (0.00005).
- I think the longitude might be a problem with the browser, since the extension doesn't edit it in any way. Please send me the names of the browsers on which it doesn't work so that I can have a look and try to find a way to fix it.
- [`enableHighAccuracy`](https://developer.mozilla.org/en-US/docs/Web/API/Geolocation/getCurrentPosition#enablehighaccuracy) is described as
> A boolean value that indicates the application would like to receive the best possible results. If true and if the device is able to provide a more accurate position, it will do so. Note that this can result in slower response times or increased power consumption (with a GPS chip on a mobile device for example). On the other hand, if false, the device can take the liberty to save resources by responding more quickly and/or using less power. Default: false.

The extension simply relays whatever the user selected to the Geolocation API.

samuellouf committed:
Changed license + fixed hat

 - Changed the license from MLP-2.0 to MIT
 - Fixed the "When the user's position changes" hat by not triggering it if the position change is inferior to 0,00005°.

samuellouf committed:
Merge branch 'TurboWarp:master' into Geolocation

samuellouf committed:
Fixed the format

samuellouf committed:
Merge branch 'Geolocation' of https://github.com/samuellouf/extensions into Geolocation

samuellouf commented:
@Brackets-Coder, could you please take a look at my Wikipedia extension (#1941)? It needs a second review. Thanks in advance!

samuellouf committed:
Merge branch 'TurboWarp:master' into Geolocation

Brackets-Coder reviewed (approved):
- [x] I verified that the extension follows all acceptance criteria and guidelines in CONTRIBUTING.md.
 - [x] I believe the block list is well-designed and intuitive.
 - [x] I tested all functionality and found no bugs, security flaws, or unintuitive behavior.
 - [x] I reviewed the code and believe it is well-written and easy to maintain or extend in the future, even without help from the original author.
 - [x] I have no additional concerns.

I think this is acceptable. Great work!
I noticed a very rare issue where the get position blocks were taking unusually long to execute, but I was unable to replicate it so I don't think it's a problem. 

Also, apparently west and east longitude are different things, and many GPS systems like Google Earth will remove the negative.... *facepalm*. There's nothing wrong with it, I just misunderstood

Even though I think this is acceptable for merging, I don't think it's perfect yet for real-time geolocation applications. I couldn't get the get position blocks to actually change value, even when walking around distances up to dozens of feet. Perhaps this is just the caching system not detecting a significant enough change? I haven't tested it while driving around and I don't think I'm going to

Brackets-Coder commented on extensions/SamuelLouf/Geolocation.js:
```diff
@@ -0,0 +1,299 @@
+
+    async isAllowed() {
+      if (!this.isSupported()) return false;
+      // @ts-ignore
+      return await Scratch.canGeolocate();
+    }
```
@SharkPool-SP this block already checks `Scratch.canGeolocate()`. Did you mean all calls to navigator.geolocate have to be wrapped in this? I didn't think so and thought it was fine

SharkPool-SP replied:
Yes

Brackets-Coder replied:
okay

Brackets-Coder replied:
@samuellouf wrap all Geolocation API calls in `Scratch.canGeolocate()` and I think you'll be good to merge. It shouldn't prompt the user except once per session

samuellouf replied:
I prefer to use this block to check if the user can geolocate. That way, if the device on which the extension is running doesn't have the geolocation API, the extension doesn't send a useless request to the user.

SharkPool-SP replied:
You may prefer it but that's not how it works here. You have to follow proper code styles.

Also, it will only prompt the user once when a block is first used, just like the microphone loudness block.

GarboMuffin replied:
I mean, this one, is not a big deal

GarboMuffin replied:
I didn't read the full context

Reading it later. Disregard that response

Brackets-Coder replied:
So @GarboMuffin should we require every block to be wrapped in canGeolocate or is it not a big deal? I'm confused

SharkPool-SP replied:
He means it should have it

GarboMuffin replied:
I said "Reading it later. Disregard that response". I did not say anything about it

samuellouf replied:
I just saw the messages and switched to "Scratch.canGeolocate()". Do any of you have any feature requests before the extension is merged?

samuellouf committed:
Merge branch 'TurboWarp:master' into Geolocation

samuellouf committed:
Switched to "Scratch.canGeolocate()"

Brackets-Coder reviewed (commented):
@samuellouf everything looks great. My only concern is that it seems like realtime geolocation isn't working for me. As stated above, this is probably a browser caching issue. Other than that, I think it's ready

SharkPool-SP commented on extensions/SamuelLouf/Geolocation.js:
```diff
@@ -0,0 +1,297 @@
+          {
+            opcode: "isSupported",
+            blockType: Scratch.BlockType.BOOLEAN,
+            text: Scratch.translate("is geolocation supported by this device?"),
```
Change this to 'device supports geolocation?'

samuellouf replied:
Done

SharkPool-SP commented on extensions/SamuelLouf/Geolocation.js:
```diff
@@ -0,0 +1,297 @@
+        util.thread._coordinates || (await getGeolocation(this.options));
+      if (coordinates.success == true) {
+        this.lastCoords = coordinates;
+        return coordinates[args.WHAT];
```
this is bad practice:

<img width="318" height="130" alt="Image" src="https://github.com/user-attachments/assets/fa7fd32a-150f-4453-a178-bc9248fb0373" />

You need to normalize it so this doesnt happen

samuellouf replied:
Done. Now it returns `""` if args.WHAT isn't in the menu (`this.getInfo().menus.coordinates.items.map((e) => e.value).includes(args.WHAT)`).

SharkPool-SP commented on extensions/SamuelLouf/Geolocation.js:
```diff
@@ -0,0 +1,297 @@
+      if (!(await Scratch.canGeolocate())) return "";
+      var coordinates =
+        util.thread._coordinates || (await getGeolocation(this.options));
+      if (coordinates.success == true) {
```
if its truthy then no need to check if its true

samuellouf replied:
Done, I think...

SharkPool-SP commented on extensions/SamuelLouf/Geolocation.js:
```diff
@@ -0,0 +1,297 @@
+    }
+
+    setTimeoutTo(args) {
+      this.options.timeout = Number(args.SECONDS) * 1000;
```
use `Scratch.Cast.toNumber`, not `Number()`

samuellouf replied:
Done

SharkPool-SP commented on extensions/SamuelLouf/Geolocation.js:
```diff
@@ -0,0 +1,297 @@
+    }
+
+    isSupported() {
+      return !!navigator.geolocation;
```
use `Scratch.Cast.toBoolean`, not `!!`

samuellouf replied:
Done

SharkPool-SP commented on extensions/SamuelLouf/Geolocation.js:
```diff
@@ -0,0 +1,297 @@
+    }
+
+    addToTimeout(args) {
+      this.options.timeout += Number(args.SECONDS) * 1000;
```
use `Scratch.Cast.toNumber`, not `Number()`

samuellouf replied:
Done

SharkPool-SP commented on extensions/SamuelLouf/Geolocation.js:
```diff
@@ -0,0 +1,297 @@
+          {
+            opcode: "isHighAccuracy",
+            blockType: Scratch.BlockType.BOOLEAN,
+            text: Scratch.translate("is the accuracy high?"),
```
Id change this to 'is high accuracy?'

samuellouf replied:
Done

SharkPool-SP commented on extensions/SamuelLouf/Geolocation.js:
```diff
@@ -0,0 +1,297 @@
+                defaultValue: 1,
+              },
+            },
+          },
```
Not a big fan of this block, given theres a set and get timeout. Theres not much of a need to get a users geolocation and change the timeout dynamically.

imo should be removed. @Brackets-Coder thoughts?

samuellouf replied:
Done

SharkPool-SP reviewed (changes requested):
Fix these and ill merge it.

Also you should improve the banner, add more contrast so its a more visible.

samuellouf committed:
Merge branch 'TurboWarp:master' into Geolocation

samuellouf committed:
Cleanup

DangoCat committed:
[Automated] Format code

SharkPool-SP committed:
Code improvements

DangoCat committed:
[Automated] Format code

samuellouf committed:
More contrast

GarboMuffin reviewed (changes requested):
I will review it and merge it. It's a sensitive enough topic

GarboMuffin committed:
clean this up. if you're not using error then why put it there?

GarboMuffin committed:
this is not an appropriate use of toBoolean

GarboMuffin committed:
more obvious errors

GarboMuffin committed:
why would you say "is TurboWarp watching" - that implies i'm getting your location. that's obviously wrong

GarboMuffin committed:
fix many block design deficiencies

GarboMuffin committed:
update description

DangoCat committed:
[Automated] Format code
